test/NuGet.Core.Tests/NuGet.Commands.Test/SignCommandRunnerTests.cs

```diff
@@ -0,0 +1,174 @@
+
+                File.WriteAllBytes(certificateFilePath, test.Certificate.Export(X509ContentType.Pkcs12, "password"));
+
+                test.Args.CertificatePath = certificateFilePath;
+                test.Args.CertificatePassword = "incorrect password";
+
+                var exception = await Assert.ThrowsAsync<SignCommandException>(
+                    () => test.Runner.ExecuteCommandAsync(test.Args));
+
+                Assert.Equal(NuGetLogCode.NU3001, exception.AsLogMessage().Code);
+                Assert.Equal($"Invalid password was provided for the certificate file '{certificateFilePath}'. Please provide a valid password using the '-CertificatePassword' option", exception.Message);
+            }
+        }
+
+        [Fact]
+        public async Task ExecuteCommandAsync_WithAmbiguousMatch_Throws()
+        {
+            using (var test = Test.Create())
+            {
+                test.Args.CertificateSubjectName = "Root";
+                test.Args.CertificateStoreLocation = StoreLocation.LocalMachine;
+                test.Args.CertificateStoreName = StoreName.Root;
+
+                var exception = await Assert.ThrowsAsync<SignCommandException>(
+                    () => test.Runner.ExecuteCommandAsync(test.Args));
+
+                Assert.Equal(NuGetLogCode.NU3001, exception.AsLogMessage().Code);
+                Assert.Equal("Multiple certificates were found that meet all the given criteria. Use the '-CertificateFingerprint' option with the hash of the desired certificate.", exception.Message);
+            }
+        }
+
+        private static byte[] GetResource(string name)
+        {
+            return ResourceTestUtility.GetResourceBytes(
+                $"NuGet.Commands.Test.compiler.resources.{name}",
+                typeof(SignCommandRunnerTests));
+        }
+
+        private sealed class Test : IDisposable
+        {
+            private bool _isDisposed;
+
+            internal SignArgs Args { get; }
+            internal X509Certificate2 Certificate { get; }
+            internal TestDirectory Directory { get; }
+            internal SignCommandRunner Runner { get; }
+
+            internal Test(SignArgs args, TestDirectory directory, X509Certificate2 certificate)
+            {
+                Args = args;
+                Directory = directory;
+                Certificate = certificate;
+                Runner = new SignCommandRunner();
+            }
+
+            public void Dispose()
+            {
+                if (!_isDisposed)
+                {
+                    Certificate.Dispose();
+
+                    GC.SuppressFinalize(this);
+
+                    _isDisposed = true;
+                }
+            }
+
+            internal static Test Create()
+            {
+                var directory = TestDirectory.Create();
+                var certificate = SigningTestUtility.GenerateCertificate("test", generator => { });
```

Comment:
refrain from using `SigningTestUtility.GenerateCertificate` per test as it is a time consuming process. Please use a test fixture to generate a code signing certificate common to all tests. For example see [SigningTestFixture](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/SigningTestFixture.cs).

Reply:
Really?  What perf numbers do you have on this?  Keep in mind that this certificate is not being installed into a certificate store.

Reply:
As part of https://github.com/NuGet/NuGet.Client/pull/1849, I did some investigation. 

At that point, the duration of running `NuGet.Packaging.Functest` locally was down from ~27 seconds to ~6 seconds by reusing the same certificate via a test fixture.